### PR TITLE
Abstract nostdinc and move xcc as standalone compiler

### LINKED ISSUE #13710: Build failure when using XCC toolchain
**Describe the bug**
With commit 9c2ac778612da408900ff73feb7310a01634dd46, XCC is invoked with `-nostdinc` flag and this causes compilation errors. The `-nostdinc` flag seems to make XCC not look for header files in `XtensaTools/xtensa-elf/include`

**To Reproduce**
1. `cd $ZEPHYR_BASE/tests/boards/intel_s1000_crb`
1. `mkdir build; cd build`
1. `cmake -DBOARD=intel_s1000_crb ..`
1. `make`
1. See error

**Expected behavior**
Build successful

**Impact**
Can no longer build using XCC

**Screenshots or console output**
```
../../../../../include/toolchain/xcc.h:21:32: error: xtensa/config/core.h: No such file or directory
../../../../../arch/xtensa/include/xtensa_context.h:26:34: error: xtensa/config/tie.h: No such file or directory
../../../../../arch/xtensa/include/xtensa_context.h:27:32: error: xtensa/corebits.h: No such file or directory
../../../../../arch/xtensa/include/xtensa_context.h:28:37: error: xtensa/config/system.h: No such file or directory
../../../../../arch/xtensa/include/xtensa_context.h:29:40: error: xtensa/xtruntime-frames.h: No such file or directory
```
**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Xtensa
 - Commit SHA or Version used: 9c2ac778612da408900ff73feb7310a01634dd46

**Additional context**
Temporarily removing `-nostdinc` flag for experimentation made the build successful.

### PR DESCRIPTION
**Pull request by nashif**
- newlib needs c standard includes (so no -nostdinc)
- xcc needs toolchain headers (so no -nostdinc)
- with host gcc:
  - x86_64 should not build with standard includes (-nostdinc needed)
  - native_posix should build with standard include (no -nostdinc)
..
..

To simplify, abstract this and move it to compilers/toolchains and still
depend on what the application wants.

Also, XCC is based on xcc, but is nothing like gcc and his many differences.
Instead of ifdeffing the gcc code with Xcc specifics, maintain it
standalone.

Fixes #13710

-------------

Part of #16031

### CODE REVIEW

**Comment by nashif** [comment]
this should resolve issue #13606 is trying to fix and other issues reported resulting from the offending change

**Review by aurel32 — APPROVED** [review]
Thanks a lot for this PR. It works as expected and fixes the issues encountered since 9c2ac778612da408900ff73feb7310a01634dd46 has been merged.

Tested with both the Debian toolchain and Zephyr SDK 0.10.0 for apps using the minimal libc, for apps using newlib and for apps using newlib + C++ headers.

My only nitpick is about the commit message of the second patch, I think there is a thinko: "for now x86_64 does not support x86_64".

File: cmake/compiler/gcc/target_nostdinc.cmake [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+macro(toolchain_cc_nostdinc)
+
+  zephyr_compile_options(
+    -nostdinc
```
Comment (mped-oticon) [inline cmake/compiler/gcc/target_nostdinc.cmake]: I'm fine with this. 
But if we are to follow precedence of the other toolchain abstractions, this should set a variable, and leave the call to `zephyr_compile_options` to the root CMakeLists.txt.

File: CMakeLists.txt [hunk 2]
```diff
@@ -187,10 +187,11 @@ if(NOT CONFIG_RTTI)
   )
 endif()
 
-if(NOT CONFIG_NATIVE_APPLICATION)
-  set(NOSTDINC_F -nostdinc)
+if (NOT CONFIG_NEWLIB_LIBC AND NOT CONFIG_NATIVE_APPLICATION)
+toolchain_cc_nostdinc()
```
Comment (mped-oticon) [inline CMakeLists.txt]: I know the name reveals the meaning here, but for consistency, we should document the intent of these macros.
nit: indent.

File: cmake/compiler/host-gcc/target_nostdinc.cmake [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
+macro(toolchain_cc_nostdinc)
+
+  zephyr_compile_options(
+    -nostdinc
+  )
+
+endmacro()
```
Comment (mped-oticon) [inline cmake/compiler/host-gcc/target_nostdinc.cmake]: If we believe this `host-gcc/target_nostdinc.cmake` file will be the same as `gcc/target_nostdinc.cmake`, then I'd vote for removing this file and let host-gcc piggy back off main gcc.

File: cmake/compiler/host-gcc/target.cmake [hunk 4]
```diff
@@ -81,3 +81,4 @@ include(${ZEPHYR_BASE}/cmake/compiler/gcc/target_security_fortify.cmake)
 include(${ZEPHYR_BASE}/cmake/compiler/gcc/target_security_canaries.cmake)
 include(${ZEPHYR_BASE}/cmake/compiler/gcc/target_optimizations.cmake)
 include(${ZEPHYR_BASE}/cmake/compiler/gcc/target_cpp.cmake)
+include(${ZEPHYR_BASE}/cmake/compiler/${COMPILER}/target_nostdinc.cmake)
```
Comment (mped-oticon) [inline cmake/compiler/host-gcc/target.cmake]: Not piggy-backing off gcc like the other includes above?

**Comment by nashif** [comment]
> My only nitpick is about the commit message of the second patch, I think there is a thinko: "for now x86_64 does not support x86_64".

heh, fixed

**Comment by mped-oticon** [comment]
A thought: The toolchain abstraction was meant to raise the abstraction level, not just wrap single flags. I can fix this later, but I'd prefer if fundamental flags like -ffreestanding and -nostdinc was part of some category. Having too low-level abstractions only solves stuff for the short term.

**Comment by nashif** [comment]
@mped-oticon I am still not happy with the changes tbh, so thanks for the feedback, I will look at naming and abstraction. this is now fixing a real bug and not just abstracting something that worked before.

**Review by sathishkuttan — APPROVED** [review]
Verified that the commits in this PR resolve #13710
LGTM

**Pull request merged**